File: dotnet/samples/agents/BasicM365Agent/LLMRecorder.cs
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.

using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using System.Security.Cryptography;
using System.Text;
using System.Text.Json;
using System.Text.Json.Serialization;
using System.Threading;
using System.Threading.Tasks;
using OpenAI.Chat;

namespace BasicM365Sample;

/// <summary>
/// Records LLM request/response pairs for test replay.
/// Based on Python's LLMRecorder implementation.
/// </summary>
public class LLMRecorder
{
    private readonly string _recordingsDir;
    private int _callCount = 0;
    private readonly JsonSerializerOptions _jsonOptions;

    public LLMRecorder(string recordingsDir)
    {
        _recordingsDir = recordingsDir ?? throw new ArgumentNullException(nameof(recordingsDir));
        Directory.CreateDirectory(_recordingsDir);

        _jsonOptions = new JsonSerializerOptions
        {
            WriteIndented = true,
            PropertyNamingPolicy = JsonNamingPolicy.CamelCase,
            DefaultIgnoreCondition = JsonIgnoreCondition.WhenWritingNull
        };
    }

    /// <summary>
    /// Generate deterministic hash from request parameters.
    /// </summary>
    public string HashRequest(
        string model,
        IEnumerable<ChatMessage> messages,
        IEnumerable<ChatTool>? tools = null,
        float temperature = 0.0f)
    {
        // Build canonical request representation
        var requestDict = new Dictionary<string, object>
        {
            ["model"] = model,
            ["messages"] = NormalizeMessages(messages),
            ["temperature"] = temperature
        };

        if (tools != null && tools.Any())
        {
            requestDict["tools"] = tools.Select(t => new
            {
                type = "function",
                function = new
                {
                    name = t.FunctionName,
                    description = t.FunctionDescription,
                    parameters = t.FunctionParameters?.ToString() ?? "{}"
                }
            }).ToArray();
        }

        // Serialize to stable JSON (sorted keys, no whitespace)
        var json = JsonSerializer.Serialize(requestDict, new JsonSerializerOptions
        {
            PropertyNamingPolicy = JsonNamingPolicy.CamelCase,
            WriteIndented = false,
            DefaultIgnoreCondition = JsonIgnoreCondition.WhenWritingNull
        });

        // Sort keys by converting to sorted dictionary
        using var doc = JsonDocument.Parse(json);
        var sortedJson = SortJsonKeys(doc.RootElement);

        // Hash and truncate
        var hash = SHA256.HashData(Encoding.UTF8.GetBytes(sortedJson));
        return Convert.ToHexString(hash)[..16].ToLowerInvariant();
    }

    /// <summary>
    /// Record an LLM request/response pair.
    /// </summary>
    public async Task RecordAsync(
        string model,
        IEnumerable<ChatMessage> messages,
        IEnumerable<ChatTool>? tools,
        ChatCompletion response,
        CancellationToken cancellationToken = default)
    {
        var callId = Interlocked.Increment(ref _callCount);
        var timestamp = DateTime.UtcNow;

        var hashKey = HashRequest(model, messages, tools);

        // Record request
        var requestData = new
        {
            callId,
            timestamp,
            hash = hashKey,
            model,
            messages = NormalizeMessages(messages),
            tools = tools?.Select(t => new
            {
                type = "function",
                function = new
                {
                    name = t.FunctionName,
                    description = t.FunctionDescription,
                    parameters = JsonSerializer.Deserialize<JsonElement>(t.FunctionParameters?.ToString() ?? "{}")
                }
            }).ToArray()
        };

        var requestFile = Path.Combine(_recordingsDir, $"{hashKey}.request.json");
        await File.WriteAllTextAsync(
            requestFile,
            JsonSerializer.Serialize(requestData, _jsonOptions),
            cancellationToken);

        // Record response
        var responseData = new
        {
            callId,
            timestamp = DateTime.UtcNow,
            hash = hashKey,
            response = new
            {
                id = response.Id,
                model = response.Model,
                created = response.CreatedAt,
                finishReason = response.FinishReason.ToString(),
                content = response.Content.Select(c => new { text = c.Text }).ToArray(),
                toolCalls = response.ToolCalls.Select(tc => new
                {
                    id = tc.Id,
                    type = tc.Kind.ToString(),
                    function = new
                    {
                        name = tc.FunctionName,
                        arguments = tc.FunctionArguments.ToString()
                    }
                }).ToArray()
            }
        };

        var responseFile = Path.Combine(_recordingsDir, $"{hashKey}.response.json");
        await File.WriteAllTextAsync(
            responseFile,
            JsonSerializer.Serialize(responseData, _jsonOptions),
            cancellationToken);
    }

    private List<object> NormalizeMessages(IEnumerable<ChatMessage> messages)
    {
        return messages.Select(msg =>
        {
            var normalized = new Dictionary<string, object>
            {
                ["role"] = msg switch
                {
                    SystemChatMessage => "system",
                    UserChatMessage => "user",
                    AssistantChatMessage => "assistant",
                    ToolChatMessage => "tool",
                    _ => "unknown"
                }
            };

            // Extract content based on message type
            if (msg is SystemChatMessage systemMsg)
            {
                normalized["content"] = systemMsg.Content[0].Text;
            }
            else if (msg is UserChatMessage userMsg)
            {
                normalized["content"] = userMsg.Content[0].Text;
            }
            else if (msg is AssistantChatMessage assistantMsg)
            {
                if (assistantMsg.Content.Any())
                {
                    normalized["content"] = string.Join("", assistantMsg.Content.Select(c => c.Text));
                }
                if (assistantMsg.ToolCalls.Any())
                {
                    normalized["tool_calls"] = assistantMsg.ToolCalls.Select(tc => new
                    {
                        id = tc.Id,
                        type = tc.Kind.ToString(),
                        function = new
                        {
                            name = tc.FunctionName,
                            arguments = tc.FunctionArguments.ToString()
                        }
                    }).ToArray();
                }
            }
            else if (msg is ToolChatMessage toolMsg)
            {
                normalized["tool_call_id"] = toolMsg.ToolCallId;
                normalized["content"] = toolMsg.Content[0].Text;
            }

            return (object)normalized;
        }).ToList();
    }

    private string SortJsonKeys(JsonElement element)
    {
        if (element.ValueKind == JsonValueKind.Object)
        {
            var sorted = new SortedDictionary<string, object>();
            foreach (var prop in element.EnumerateObject())
            {
                if (prop.Value.ValueKind == JsonValueKind.Object || prop.Value.ValueKind == JsonValueKind.Array)
                {
                    sorted[prop.Name] = JsonSerializer.Deserialize<object>(SortJsonKeys(prop.Value))!;
                }
                else
                {
                    sorted[prop.Name] = JsonSerializer.Deserialize<object>(prop.Value.GetRawText())!;
                }
            }
            return JsonSerializer.Serialize(sorted);
        }
        else if (element.ValueKind == JsonValueKind.Array)
        {
            var array = element.EnumerateArray()
                .Select(item => JsonSerializer.Deserialize<object>(
                    item.ValueKind == JsonValueKind.Object || item.ValueKind == JsonValueKind.Array
                        ? SortJsonKeys(item)
                        : item.GetRawText())!)
                .ToArray();
            return JsonSerializer.Serialize(array);
        }

        return element.GetRawText();
    }
}

File: typescript/packages/agents-protocol-hosting/src/llm/llm-recorder.ts
import * as fs from 'fs/promises';
import * as crypto from 'crypto';
import * as path from 'path';

export interface LLMRequest {
  model: string;
  messages: Array<{
    role: string;
    content: string;
    tool_calls?: any[];
    tool_call_id?: string;
  }>;
  tools?: Array<{
    type: string;
    function: {
      name: string;
      description?: string;
      parameters?: any;
    };
  }>;
  temperature?: number;
}

export interface LLMResponse {
  id: string;
  model: string;
  created: number;
  choices: Array<{
    finish_reason: string;
    message: {
      role: string;
      content: string | null;
      tool_calls?: Array<{
        id: string;
        type: string;
        function: {
          name: string;
          arguments: string;
        };
      }>;
    };
  }>;
  usage?: {
    prompt_tokens?: number;
    completion_tokens?: number;
    total_tokens?: number;
  };
}

/**
 * Records LLM request/response pairs for deterministic testing.
 * Based on the .NET LLMRecorder implementation.
 */
export class LLMRecorder {
  private callCount = 0;

  constructor(private recordingsDir: string) {}

  /**
   * Generate deterministic hash from request parameters.
   */
  hashRequest(request: LLMRequest): string {
    // Build canonical request representation
    const requestDict: Record<string, any> = {
      model: request.model,
      messages: this.normalizeMessages(request.messages),
      temperature: request.temperature || 0.0
    };

    if (request.tools && request.tools.length > 0) {
      requestDict.tools = request.tools.map(t => ({
        type: t.type,
        function: {
          name: t.function.name,
          description: t.function.description,
          parameters: t.function.parameters
        }
      }));
    }

    // Serialize to stable JSON (sorted keys)
    const json = this.sortedStringify(requestDict);

    // Hash and truncate
    const hash = crypto.createHash('sha256').update(json).digest('hex');
    return hash.substring(0, 16);
  }

  /**
   * Record an LLM request/response pair.
   */
  async recordAsync(request: LLMRequest, response: LLMResponse): Promise<void> {
    const callId = ++this.callCount;
    const timestamp = new Date().toISOString();
    const hashKey = this.hashRequest(request);

    // Ensure directory exists
    await fs.mkdir(this.recordingsDir, { recursive: true });

    // Record request
    const requestData = {
      callId,
      timestamp,
      hash: hashKey,
      model: request.model,
      messages: this.normalizeMessages(request.messages),
      tools: request.tools
    };

    const requestFile = path.join(this.recordingsDir, `${hashKey}.request.json`);
    await fs.writeFile(requestFile, JSON.stringify(requestData, null, 2));

    // Record response
    const responseData = {
      callId,
      timestamp: new Date().toISOString(),
      hash: hashKey,
      response: {
        id: response.id,
        model: response.model,
        created: response.created,
        finishReason: response.choices[0]?.finish_reason || 'stop',
        content: response.choices[0]?.message.content
          ? [{ text: response.choices[0].message.content }]
          : [],
        toolCalls: response.choices[0]?.message.tool_calls?.map(tc => ({
          id: tc.id,
          type: tc.type,
          function: {
            name: tc.function.name,
            arguments: tc.function.arguments
          }
        })) || []
      }
    };

    const responseFile = path.join(this.recordingsDir, `${hashKey}.response.json`);
    await fs.writeFile(responseFile, JSON.stringify(responseData, null, 2));

    console.log(`  🔴 Recorded LLM call #${callId}: ${hashKey}`);
  }

  private normalizeMessages(messages: Array<any>): Array<any> {
    return messages.map(msg => {
      const normalized: any = {
        role: msg.role
      };

      if (msg.content) {
        normalized.content = msg.content;
      }

      if (msg.tool_calls) {
        normalized.tool_calls = msg.tool_calls.map((tc: any) => ({
          id: tc.id,
          type: tc.type,
          function: {
            name: tc.function.name,
            arguments: tc.function.arguments
          }
        }));
      }

      if (msg.tool_call_id) {
        normalized.tool_call_id = msg.tool_call_id;
      }

      return normalized;
    });
  }

  private sortedStringify(obj: any): string {
    if (Array.isArray(obj)) {
      return '[' + obj.map(item => this.sortedStringify(item)).join(',') + ']';
    } else if (obj !== null && typeof obj === 'object') {
      const keys = Object.keys(obj).sort();
      const pairs = keys.map(key => `"${key}":${this.sortedStringify(obj[key])}`);
      return '{' + pairs.join(',') + '}';
    } else {
      return JSON.stringify(obj);
    }
  }
}
